test_rclcpp/test/test_executor.cpp

```diff
@@ -118,6 +121,84 @@ TEST(CLASSNAME(test_executor, RMW_IMPLEMENTATION), multiple_executors) {
   ASSERT_THROW(executor2.add_node(node1), std::runtime_error);
 }
 
+// Check that the executor is notified when a node adds a new timer, publisher, subscription,
+// service or client.
+TEST(CLASSNAME(test_executor, RMW_IMPLEMENTATION), notify) {
+  rclcpp::executors::SingleThreadedExecutor executor;
+  auto executor_spin_lambda = [&executor]() {
+      executor.spin();
+    };
+  auto node = rclcpp::Node::make_shared("test_executor_notify");
+  executor.add_node(node);
+  {
+    std::thread spin_thread(executor_spin_lambda);
+    bool timer_triggered = false;
+    auto timer = node->create_wall_timer(
+      1_ms,
+      [&executor, &timer_triggered](rclcpp::TimerBase & timer)
+    {
+      timer_triggered = true;
+      executor.cancel();
+      timer.cancel();
+    });
+
+    spin_thread.join();
+    EXPECT_TRUE(timer_triggered);
+  }
+
+  {
+    std::thread spin_thread(executor_spin_lambda);
+    bool subscription_triggered = false;
+    auto sub_callback =
+      [&executor, &subscription_triggered](test_rclcpp::msg::UInt32::ConstSharedPtr msg) -> void
+      {
+        subscription_triggered = true;
+        EXPECT_EQ(msg->data, 42);
+        executor.cancel();
+      };
+    auto subscription = node->create_subscription<test_rclcpp::msg::UInt32>(
+      "test_executor_notify_subscription",
+      sub_callback,
+      rmw_qos_profile_default);
+
+    auto publisher = node->create_publisher<test_rclcpp::msg::UInt32>(
+      "test_executor_notify_subscription", rmw_qos_profile_default);
+    test_rclcpp::msg::UInt32 pub_msg;
+    pub_msg.data = 42;
+    publisher->publish(pub_msg);
+    spin_thread.join();
+
+    EXPECT_TRUE(subscription_triggered);
+  }
+
+  {
+    std::thread spin_thread(executor_spin_lambda);
+
+    auto service = node->create_service<test_rclcpp::srv::AddTwoInts>(
+      "test_executor_notify_service",
+      [](test_rclcpp::srv::AddTwoInts::Request::SharedPtr request,
+      test_rclcpp::srv::AddTwoInts::Response::SharedPtr response)
+    {
+      response->sum = request->a + request->b;
+    });
+
+    auto client = node->create_client<test_rclcpp::srv::AddTwoInts>(
+      "test_executor_notify_service"
+      );
+    auto request = std::make_shared<test_rclcpp::srv::AddTwoInts::Request>();
+    request->a = 4;
+    request->b = 2;
+    auto future_result = client->async_send_request(request);
+    EXPECT_EQ(future_result.get()->sum, 6);
+    executor.cancel();
+    spin_thread.join();
+  }
+}
+
+// test removing a node
+
+// test notify with multiple nodes
```

Comment:
I have similarly merged things before I had what I considered complete coverage. I would say it's up to you. You can just put in TODO's / make an issue to track them (so we don't forget about them) or you can try to implement them now (maybe you have a clearer picture of how to do so right now rather than in the future).